subprojects/scheduler/src/main/java/org/gradle/internal/scheduler/Graph.java

```diff
@@ -0,0 +1,291 @@
+            dag.addEdge(edge);
+            parents.put(target, source);
+        }
+
+        // Then add any breakable edges as long as they don't form a cycle
+        for (Edge removableEdge : removableEdges) {
+            Node source = removableEdge.getSource();
+            Node target = removableEdge.getTarget();
+            if (hasAmongAncestors(parents, source, target)) {
+                // Skip edge if it would form a cycle
+                continue;
+            }
+            dag.addEdge(removableEdge);
+            // TODO Need to track all weak parents, not just one
+            parents.put(target, source);
+        }
+
+        return dag;
+    }
+
+    private static void addAncestorsToPath(Map<Node, Node> parents, Node child, List<Node> path) {
+        path.add(child);
+        Node parent = parents.get(child);
+        if (parent != null) {
+            addAncestorsToPath(parents, parent, path);
+        }
+    }
+
+    private static boolean hasAmongAncestors(Map<Node, Node> parents, Node parent, Node child) {
+        Node grandParent = parents.get(parent);
+        if (grandParent == null) {
+            return false;
+        }
+        if (grandParent == child) {
+            return true;
+        }
+        return hasAmongAncestors(parents, grandParent, child);
+    }
+
+    public void walkIncomingEdgesFrom(Node start, EdgeWalkerAction action) {
+        Deque<Node> queue = new ArrayDeque<Node>();
+        queue.add(start);
+        while (true) {
+            Node node = queue.poll();
+            if (node == null) {
+                break;
+            }
+            for (Edge incoming : Lists.newArrayList(incomingEdges.get(node))) {
+                if (action.execute(incoming)) {
+                    queue.add(incoming.getSource());
+                }
+            }
+        }
+    }
+
+    public interface EdgeWalkerAction {
+        boolean execute(Edge edge);
+    }
+
+    /**
+     * Removes nodes with their incoming edges that are not accessible from the given entry nodes via live edges.
+     */
+    public void removeDeadNodes(Collection<? extends Node> entryNodes, LiveEdgeDetector detector) {
+        Set<Node> liveNodes = Sets.newLinkedHashSet();
+        Deque<Node> queue = new ArrayDeque<Node>(entryNodes);
+        while (true) {
+            Node node = queue.poll();
+            if (node == null) {
+                break;
+            }
+            if (!liveNodes.add(node)) {
+                continue;
+            }
+            for (Edge incoming : incomingEdges.get(node)) {
+                Node source = incoming.getSource();
+                if (!liveNodes.contains(source) && detector.isIncomingEdgeLive(incoming)) {
+                    queue.add(source);
+                }
+            }
+            for (Edge outgoing : outgoingEdges.get(node)) {
+                Node target = outgoing.getTarget();
+                if (!liveNodes.contains(target) && detector.isOutgoingEdgeLive(outgoing)) {
```

Comment:
We could remove the `contains` check here since we already do it a little bit further up for every node coming from the queue.

Reply:
I meant the
```
if (!liveNodes.add(node)) {
    continue;
}
```
check. That does not happen in the current round but as soon as it has been added and then removed from the queue.

Reply:
Ah, got it. Fixed.